### PR TITLE
Fix Point2D crash under evaluate(False) (swev-id: sympy__sympy-22714)

### LINKED ISSUE #139: Point2D: sympify crashes under with evaluate(False) — Imaginary coordinates error
## User request

simplicit gives `Imaginary coordinates are not permitted.` with evaluate(False)

## Issue
`with evaluate(False)` crashes unexpectedly with `Point2D`

## Code
```python
import sympy as sp
with sp.evaluate(False):
  sp.S('Point2D(Integer(1),Integer(2))')
```

## Error
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/core/sympify.py", line 472, in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 1026, in parse_expr
    raise e from ValueError(f"Error from parse_expr with transformed code: {code!r}")
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 1017, in parse_expr
    rv = eval_expr(code, local_dict, global_dict)
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 911, in eval_expr
    expr = eval(
  File "<string>", line 1, in <module>
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/geometry/point.py", line 912, in __new__
    args = Point(*args, **kwargs)
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/geometry/point.py", line 153, in __new__
    raise ValueError('Imaginary coordinates are not permitted.')
ValueError: Imaginary coordinates are not permitted.
```

However, it works without `with evaluate(False)`. Both of following commands work
```python
sp.S('Point2D(Integer(1),Integer(2))')
sp.S('Point2D(Integer(1),Integer(2))', evaluate=False)
```

## Specification (research)
Root cause:
- In `sympy/geometry/point.py`, `Point.__new__` checks imaginary coordinates with `any(a.is_number and im(a) for a in coords)`. Under `with evaluate(False)`, `im(a)` can remain unevaluated (e.g., `im(1)`), and its truthiness erroneously evaluates to True, raising `ValueError` for real inputs.

Proposed fix:
- Replace boolean truthiness of `im(a)` with assumption-based check:
  - Change to: `if any(a.is_number and (a.is_real is False) for a in coords): raise ValueError(...)`
  - Alternatively: `if any(a.is_number and (im(a).is_zero is False) for a in coords): ...`
- This preserves behavior while making the check robust to `evaluate(False)`.

Tests:
- Add tests in `sympy/geometry/tests/test_point.py`:
  - With `evaluate(False)` context: `S('Point2D(Integer(1),Integer(2))') == Point2D(1, 2)`.
  - Ensure complex numeric still fails: `raises(ValueError, lambda: S('Point2D(Integer(1), I)'))`.
  - Flag `evaluate=False` directly: `S('Point2D(Integer(1),Integer(2))', evaluate=False) == Point2D(1, 2)`.
  - Without context: same success case.

Compatibility:
- No API change; real numeric coordinates allowed; complex numeric coordinates still disallowed. Fix only avoids the incorrect crash when the context suppresses evaluation.

### PR DESCRIPTION
**Pull request by rowan-stein**
## Summary
Fixes a bug where constructing a Point2D via `S('...')` crashes under `with evaluate(False)` with `ValueError: Imaginary coordinates are not permitted.`

- Root cause: In `Point.__new__`, the check `any(a.is_number and im(a) for a in coords)` relies on the boolean truthiness of `im(a)`. Under `with evaluate(False)`, `im(a)` may remain unevaluated (e.g., `im(1)`), which is truthy, incorrectly triggering the ValueError for real numeric inputs.
- Fix: Replace the check with an assumptions-based guard: `any(a.is_number and (a.is_real is False) for a in coords)`, preserving intended behavior while avoiding boolean truthiness on unevaluated expressions.

Closes #139

## Reproduction steps
```python
import sympy as sp
with sp.evaluate(False):
  sp.S('Point2D(Integer(1),Integer(2))')
```

## Observed failure
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/core/sympify.py", line 472, in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 1026, in parse_expr
    raise e from ValueError(f"Error from parse_expr with transformed code: {code!r}")
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 1017, in parse_expr
    rv = eval_expr(code, local_dict, global_dict)
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 911, in eval_expr
    expr = eval(
  File "<string>", line 1, in <module>
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/geometry/point.py", line 912, in __new__
    args = Point(*args, **kwargs)
  File "/home/avinash/.local/lib/python3.8/site-packages/sympy/geometry/point.py", line 153, in __new__
    raise ValueError('Imaginary coordinates are not permitted.')
ValueError: Imaginary coordinates are not permitted.
```

## Verification
- Without context: `S('Point2D(Integer(1),Integer(2))')` returns `Point2D(1, 2)`.
- With flag: `S('Point2D(Integer(1),Integer(2))', evaluate=False)` returns `Point2D(1, 2)`.
- With context manager after this fix: `Point2D(1, 2)` is constructed successfully.

## Tests
- Added regression tests in `sympy/geometry/tests/test_point.py` covering:
  - `with evaluate(False)` for the `S('Point2D(Integer(1),Integer(2))')` case.
  - Complex coordinates remain disallowed, both with and without the context.
  - Existing successful cases remain intact.

## Notes
- CI intentionally not triggered per task instructions.

### CODE REVIEW

File: sympy/geometry/point.py [hunk 1]
```diff
@@ -152,7 +151,7 @@ def __new__(cls, *args, **kwargs):
                         'warn' or 'ignore'.'''))
         if any(coords[dim:]):
             raise ValueError('Nonzero coordinates cannot be removed.')
-        if any(a.is_number and im(a) for a in coords):
+        if any(a.is_number and (a.is_real is False) for a in coords):
```
Comment (noa-lucent) [inline sympy/geometry/point.py]: [major] This guard now only rejects coordinates that SymPy can *prove* satisfy . Unfortunately a large class of numeric expressions with no free symbols (e.g. , , etc.) have  return  even though they are genuinely non-real. With this change those inputs slip through and  will accept complex numeric coordinates, which violates the invariant enforced by this block. Before, the  check rejected them because the imaginary part remained a nonzero expression. Please keep the invariant by handling the  case (for example, fall back to an imaginary-part test that avoids relying on truthiness) and add a regression test that covers a complex numeric where  is .}

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for digging into the `evaluate(False)` regression. The guard you replaced now only fires when `is_real` is **explicitly** `False`. Many numeric expressions that are genuinely complex still report `None` for that property (e.g. `gamma(1 + I)`), so they would now be accepted as coordinates, breaking the invariant this block enforces. Please handle the `None` case (for example by falling back to an imaginary-part check that avoids boolean truthiness) and add a regression test covering such a complex numeric.

**Comment by casey-brooks** [comment]
Test & Lint Summary
- PATH=/root/.local/bin:$PATH bin/test sympy/geometry/tests/test_point.py → 13 passed / 0 failed / 0 skipped
- PATH=/root/.local/bin:$PATH bin/test code_quality → 7 passed / 0 failed / 0 skipped

File: sympy/geometry/point.py [hunk 2]
```diff
@@ -152,7 +152,8 @@ def __new__(cls, *args, **kwargs):
                         'warn' or 'ignore'.'''))
         if any(coords[dim:]):
             raise ValueError('Nonzero coordinates cannot be removed.')
```
Comment (noa-lucent) [inline sympy/geometry/point.py:154]: [major] Thanks for tightening the check, but this still lets genuinely complex numerics slip through. Consider `Point2D(E**I, 0)`: `E**I` has no free symbols so `is_number` is true, `is_real` is `None`, and `im(E**I)` simplifies to `E*sin(1)`. For that expression `im(...).is_zero` is also `None`, so the guard does not fire and the point is constructed successfully even though the coordinate is non-real. We still need to block numbers whose imaginary part cannot be proven zero.

Could we fall back to a semantic check such as `im(a).equals(0)` (or any other robust predicate) when `is_zero` returns `None`? That keeps the regression fix for real inputs under `evaluate(False)` while rejecting complex numerics like `E**I`. Please also add a regression test that exercises one of these `is_zero -> None` cases.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
The updated check is closer, but it still accepts complex numerics like `E**I` because both `is_real` and `im(...).is_zero` return `None`. We still need to block those cases and add a regression test covering them.

**Comment by casey-brooks** [comment]
Test & Lint Summary
- PATH=/root/.local/bin:$PATH bin/test sympy/geometry/tests/test_point.py → 13 passed / 0 failed / 0 skipped
- PATH=/root/.local/bin:$PATH bin/test code_quality → 7 passed / 0 failed / 0 skipped

**Review by noa-lucent — APPROVED** [review]
Thanks for the quick follow-up. The guard now handles the uncertain cases correctly, and the new regression test covers the `E**I` scenario. Everything looks good to me.
